solidity/contracts/l2/L2WormholeGateway.sol

```diff
@@ -0,0 +1,275 @@
+            "Bridge token address must not be 0x0"
+        );
+        require(
+            address(_tbtc) != address(0),
+            "L2TBTC token address must not be 0x0"
+        );
+
+        bridge = _bridge;
+        bridgeToken = _bridgeToken;
+        tbtc = _tbtc;
+        mintingLimit = type(uint256).max;
+    }
+
+    /// @notice This function is called when the user redeems their token on L2.
+    ///         The contract receives Wormhole tBTC representation and mints the
+    ///         canonical tBTC for the user.
+    ///         If the tBTC minting limit has been reached by this contract,
+    ///         instead of minting tBTC the receiver address receives Wormhole
+    ///         tBTC representation.
+    /// @dev Requirements:
+    ///      - The receiver of Wormhole tBTC should be the L2WormholeGateway
+    ///        contract.
+    ///      - The receiver of the canonical tBTC should be abi-encoded in the
+    ///        payload.
+    ///      - The receiver of the canonical tBTC must not be the zero address.
+    /// @param encodedVm A byte array containing a Wormhole VAA signed by the
+    ///        guardians.
+    function receiveWormhole(bytes memory encodedVm) external nonReentrant {
+        // ITokenBridge.completeTransferWithPayload completes a contract-controlled
+        // transfer of an ERC20 token. Calling this function is not enough to
+        // ensure L2WormholeGateway received Wormhole tBTC representation.
+        // Instead of going too deep into the ITokenBridge implementation,
+        // asserting who is the receiver of the token, and which token it is,
+        // we check the balance before the ITokenBridge call and the balance
+        // after ITokenBridge call. This way, we are sure this contract received
+        // Wormhole tBTC token in the given amount. This is transparent to
+        // all potential upgrades of ITokenBridge implementation and no other
+        // validations are needed.
+        uint256 balanceBefore = bridgeToken.balanceOf(address(this));
+        bytes memory encoded = bridge.completeTransferWithPayload(encodedVm);
+        uint256 balanceAfter = bridgeToken.balanceOf(address(this));
+
+        uint256 amount = balanceAfter - balanceBefore;
+        address receiver = abi.decode(
+            bridge.parseTransferWithPayload(encoded).payload,
+            (address)
+        );
+        require(receiver != address(0), "0x0 receiver not allowed");
+
+        // We send wormhole tBTC OR mint canonical tBTC. We do not want to send
+        // dust. Sending wormhole tBTC is an exceptional situation and we want
+        // to keep it simple.
+        if (mintedAmount + amount > mintingLimit) {
+            bridgeToken.safeTransfer(receiver, amount);
+        } else {
+            // Function is nonReentrant and we need to update the balance with
+            // a call to completeTransferWithPayload first to know the amount.
```

Comment:
stale comment?

Reply:
No, this is an explanation of why do we `// slither-disable-next-line reentrancy-benign` one line later 😉 

Slither complains:
```
Reentrancy in L2WormholeGateway.receiveWormhole(bytes) (contracts/l2/L2WormholeGateway.sol#172-210):
        External calls:
        - encoded = bridge.completeTransferWithPayload(encodedVm) (contracts/l2/L2WormholeGateway.sol#184)
        State variables written after the call(s):
        - mintedAmount += amount (contracts/l2/L2WormholeGateway.sol#202)
Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#reentrancy-vulnerabilities-2
```